src/librustc_borrowck/borrowck/mod.rs

```diff
@@ -89,6 +90,8 @@ pub struct AnalysisData<'a, 'tcx: 'a> {
 fn borrowck<'a, 'tcx>(tcx: TyCtxt<'a, 'tcx, 'tcx>, owner_def_id: DefId)
     -> Lrc<BorrowCheckResult>
 {
+    assert!(tcx.borrowck_mode() != BorrowckMode::Mir);
```

Comment:
Nit: I would do 

```rust
assert!(match tcx.borrowck_mode() {
    BorrowckMode::Mir => false,
    BorrowckMode::Ast => true,
    // etc
});
```

i.e., spell out all the cases. That way, when we change the enum (as we plan to soon) we can be sure it is all correct.